pkg/hostpath/hostpath.go

```diff
@@ -163,52 +177,195 @@ func commonUpsertRanger(podPath string, key, value interface{}) bool {
 	return true
 }
 
-func commonDeleteRanger(podPath string, key interface{}) bool {
+func commonDeleteRanger(podPath, filter string, key interface{}) bool {
+	if key != filter {
+		return true
+	}
 	podFilePath := filepath.Join(podPath, fmt.Sprintf("%s", key))
 	os.Remove(podFilePath)
 	return true
 }
 
-func (hp *hostPath) mapVolumeToPod(hpv *hostPathVolume) error {
-	podConfigMapsPath := filepath.Join(hpv.TargetPath, "configmaps")
+func shareDeleteRanger(hp *hostPath, key interface{}) bool {
+	shareId := key.(string)
+	targetPath := ""
+	volID := ""
+	for _, hpv := range hostPathVolumes {
+		if hpv.SharedDataId == shareId {
+			switch hpv.SharedDataKind {
+			case "ConfigMap":
+				targetPath = filepath.Join(hpv.TargetPath, "configmaps")
+			case "Secret":
+				targetPath = filepath.Join(hpv.TargetPath, "secrets")
+			}
+			volID = hpv.VolID
+			// deleting the share effectively deletes permission to the
+			// data so we set the allowed bit to false; this will have bearing
+			// if the share is added again at a later date and the associated
+			// pod in question is still up
+			hpv.Allowed = false
+			break
+		}
+	}
+	if len(volID) > 0 && len(targetPath) > 0 {
+		err := os.RemoveAll(targetPath)
+		if err != nil {
+			klog.Warningf("share %s vol %s target path %s delete error %s",
+				shareId, volID, targetPath, err.Error())
+		}
+		// we just delete the associated data from the previously provisioned volume;
+		// we don't delete the volume in case the share is added back
+	}
+	return true
+}
+
+func shareUpdateRanger(key, value interface{}) bool {
+	shareId := key.(string)
+	share := value.(*sharev1alpha1.Share)
+	klog.V(4).Infof("share update ranger id %s share name %s", shareId, share.Name)
+	oldTargetPath := ""
+	volID := ""
+	change := false
+	lostPermissions := false
+	gainedPermissions := false
+	hpv := &hostPathVolume{}
+	for _, hpv = range hostPathVolumes {
+		if hpv.SharedDataId == shareId {
+			klog.V(4).Infof("share update ranger id %s found volume %s", shareId, hpv.VolID)
+			a, err := client.ExecuteSAR(shareId, hpv.PodNamespace, hpv.PodName, hpv.PodSA)
```

Comment:
So this is effectively how we check if a pod still has permissions for its share.  Since it is based on the share update, it works off of the share relist interval, so it is a "lazy check".

I did this vs. watching all the ClusterRoleBindings and RoleBindings and then deciphering which were related to the shares currently being consumed by Pods.

Making the relist interval for the share watch configurable has occurred to me as a way to accelerate things.  But I have not added it as of yet.

Seems like it should be safe, given we should not expect an abundance of this per se, but I'm interested in opinions.

Reply:
The share relist interval is now configurable as a container arg